### ISSUE TITLE
declare -i is silently ignored

### ISSUE THREAD

**Issue by meithecatte** [oils-for-unix/oils#2072]
```bash
$ osh -c 'declare -i x=2; x+=2; declare -p x'
declare -- x=22
$ zsh -c 'declare -i x=2; x+=2; declare -p x'
typeset -i x=4
$ bash -c 'declare -i x=2; x+=2; declare -p x'
declare -i x="4"
```

**Comment by andychu** [comment]
Oh actually this is changed for Oils 0.23.0 - @akinomyoga ran into this too

```
$ bin/osh -c 'declare -i x=2'
  declare -i x=2
  ^~~~~~~
[ -c flag ]:1: 'declare' doesn't implement flag -i (shopt --set ignore_flags_not_impl)
```

So at least it is not ignored, but it is not implemented 

---

If you run into the need for `-i`, I'd be curious where it  came up

It definitely came up in Nix and ble.sh

**Comment by andychu** [comment]
Also here is a relevant Zulip thread

https://oilshell.zulipchat.com/#narrow/stream/384942-language-design/topic/Fundamental.20reasons.20for.20OSH.20incompatibility

In the past, our attitude was more like "well `declare -i` adds no power over posix, it's just syntax" -- `x=$(( x + 2 ))` works everywhere

But now that we have YSH "figured out", OSH is more free to be more compatible

**Comment by meithecatte** [comment]
Ah, I'd say that deciding not to implement this is a valid approach. Sorry for not making sure to check the latest release – turns out that the Arch repos are outdated. Sorry for that, it's really out of character for them.

I don't necessarily *need* `declare -i`. I just heard of Oil being largely compatible with bash, and I decided to check it out by running [my implementation of crypto primitives](https://github.com/meithecatte/bassh/). X25519 and AES pass their tests, but SHA256 fails – this is what I tracked down the issue to.

When I learned of bash's `declare -i` function, I assumed that it has some performance benefits, and so I chose to use it everywhere I could, and then ended up depending on the behavior as I discovered more of its properties.

TL;DR: I am clinically insane and you can freely disregard me, as Jesus weeps with every line of bash I write.

**Comment by andychu** [comment]
Oh you should put that on https://github.com/oils-for-unix/oils/wiki/Shell-Programs-That-Run-Under-OSH  !  

We want to show people that lots of complex shell programs work!  Lots of people don't know that

---

Also how fast is it under OSH vs. bash?


---

If you are good with bash, I am going to do a bug bounty 

e.g. this pure bash program hangs under OSH, but I don't know why

https://lobste.rs/s/cjbaxu/ps_aux_written_entirely_bash_without_ever

https://oilshell.zulipchat.com/#narrow/stream/417617-help-wanted/topic/psaux.2Ebash.20doesn't.20run.20under.20OSH


@ellen364 was recently award a bug bounty for finding a bug in our JSON parser

So I want to do similar things for "figuring out how to make real shell programs run"

all of these count:

- diagnosing / reducing the bug
- adding failing tests
- fixing it

**Comment by meithecatte** [comment]
> Oh you should put that on https://github.com/oils-for-unix/oils/wiki/Shell-Programs-That-Run-Under-OSH !
> 
> We want to show people that lots of complex shell programs work! Lots of people don't know that

Well, with the caveat that it needs `-O ignore_flags_not_impl` in OSH 0.23.

> Also how fast is it under OSH vs. bash?

For AES, OSH is 80% to 40% slower depending on the key size.

```
~/dev/bassh$ RUN_TESTS=1 bash aes.sh
Testing AES...
AES-128 (x1000)

real	0m0.866s
user	0m0.861s
sys	0m0.000s
AES-192 (x1000)

real	0m1.054s
user	0m1.047s
sys	0m0.000s
AES-256 (x1000)

real	0m1.246s
user	0m1.237s
sys	0m0.001s
~/dev/bassh$ RUN_TESTS=1 ~/src/oils-for-unix-0.23.0/_bin/cxx-opt-sh/osh -O ignore_flags_not_impl aes.sh
Testing AES...
AES-128 (x1000)
real	1.576
user	1.560
sys	0.002
AES-192 (x1000)
real	1.636
user	1.622
sys	0.002
AES-256 (x1000)
real	1.794
user	1.777
sys	0.002
```

This is after I optimized with bash as the benchmark, that is. I tried running an earlier version and that one seems to actually rely on the `-i`:

```
          printf "%02x" $c
                 ^
tests.sh:11: printf expected an integer, got 'aes_block[5]'
```

X25519 comes out at 60% slower than bash. I wonder if inlining things makes as much as an impact as it does for bash...

**Comment by meithecatte** [comment]
> If you are good with bash, I am going to do a bug bounty
>
> e.g. this pure bash program hangs under OSH, but I don't know why
>
> https://lobste.rs/s/cjbaxu/ps_aux_written_entirely_bash_without_ever
>
> https://oilshell.zulipchat.com/#narrow/stream/417617-help-wanted/topic/psaux.2Ebash.20doesn't.20run.20under.20OSH

That specific thing is because you're not implementing `read`'s `-u` option. Not sure if that counts.

**Comment by meithecatte** [comment]
I took a look at the code, and I feel like silently ignoring unimplemented (or typo'ed) options is a systemic issue. Any plans on fixing it once and for good?

**Comment by andychu** [comment]
Yes, this counts!  Thank you for finding that :)

And yes it is bad that we ignore `-u` !  I didn't realize that

Looks like `read -p` is also a no-op, which is bad

I think I added that a long time ago to "get past" some error, but it's clearly not serving us well now


---

Another thing I noticed is last release cycle is that we also ignore many `shopt`, instead of giving an error


Related thread - https://oilshell.zulipchat.com/#narrow/stream/384942-language-design/topic/We.20should.20not.20ignore.20unimplemented.20flags.2C.20shopt

**Comment by andychu** [comment]
I still have the ignored flags/shopt change in mind, but this `declare -i` issue was fixed

Thanks for the report!
